libfwupdplugin/fu-context.c

```diff
@@ -1498,6 +1539,179 @@ fu_context_get_esp_volumes(FuContext *self, GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_pefile_firmware_new();
+	g_autoptr(GFile) file = g_file_new_for_path(filename);
+	fu_firmware_set_filename(firmware, filename);
+	if (!fu_firmware_parse_file(firmware, file, FWUPD_INSTALL_FLAG_NONE, error)) {
+		g_prefix_error(error, "failed to load %s: ", filename);
+		return NULL;
+	}
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_context_get_esp_files_for_entry(FuContext *self,
+				   FuEfiLoadOption *entry,
+				   GPtrArray *files,
+				   FuContextEspFileFlags flags,
+				   GError **error)
+{
+	g_autofree gchar *dp_filename = NULL;
+	g_autofree gchar *filename = NULL;
+	g_autofree gchar *mount_point = NULL;
+	g_autoptr(FuDeviceLocker) volume_locker = NULL;
+	g_autoptr(FuEfiFilePathDevicePath) dp_path = NULL;
+	g_autoptr(FuEfiHardDriveDevicePath) dp_hdd = NULL;
+	g_autoptr(FuFirmware) dp_list = NULL;
+	g_autoptr(FuVolume) volume = NULL;
+
+	/* all entries should have a list */
+	dp_list =
+	    fu_firmware_get_image_by_gtype(FU_FIRMWARE(entry), FU_TYPE_EFI_DEVICE_PATH_LIST, NULL);
+	if (dp_list == NULL)
+		return TRUE;
+
+	/* HDD */
+	dp_hdd = FU_EFI_HARD_DRIVE_DEVICE_PATH(
+	    fu_firmware_get_image_by_gtype(FU_FIRMWARE(dp_list),
+					   FU_TYPE_EFI_HARD_DRIVE_DEVICE_PATH,
+					   NULL));
+	if (dp_hdd == NULL)
+		return TRUE;
+
+	/* FILE */
+	dp_path = FU_EFI_FILE_PATH_DEVICE_PATH(
+	    fu_firmware_get_image_by_gtype(FU_FIRMWARE(dp_list),
+					   FU_TYPE_EFI_FILE_PATH_DEVICE_PATH,
+					   NULL));
+	if (dp_path == NULL)
+		return TRUE;
+
+	/* can we match the volume? */
+	volume = fu_context_get_esp_volume_by_hard_drive_device_path(self, dp_hdd, error);
+	if (volume == NULL)
+		return FALSE;
+	volume_locker = fu_volume_locker(volume, error);
+	if (volume_locker == NULL)
+		return FALSE;
+	dp_filename = fu_efi_file_path_device_path_get_name(dp_path, error);
+	if (dp_filename == NULL)
+		return FALSE;
+
+	/* the file itself */
+	mount_point = fu_volume_get_mount_point(volume);
+	filename = g_build_filename(mount_point, dp_filename, NULL);
+	if (flags & FU_CONTEXT_ESP_FILE_FLAG_INCLUDE_FIRST_STAGE) {
+		g_autoptr(FuFirmware) firmware = fu_context_esp_load_pe_file(filename, error);
+		if (firmware == NULL)
+			return FALSE;
+		fu_firmware_set_idx(firmware, fu_firmware_get_idx(FU_FIRMWARE(entry)));
+		g_ptr_array_add(files, g_steal_pointer(&firmware));
+	}
+
+	/* the 2nd stage bootloader, typically grub */
+	if (flags & FU_CONTEXT_ESP_FILE_FLAG_INCLUDE_SECOND_STAGE &&
+	    g_pattern_match_simple("*shim*.efi", filename)) {
+		g_autoptr(GString) filename2 = g_string_new(filename);
+		g_string_replace(filename2, "shim", "grub", 1);
+		if (g_file_test(filename2->str, G_FILE_TEST_EXISTS)) {
+			g_autoptr(FuFirmware) firmware =
+			    fu_context_esp_load_pe_file(filename2->str, error);
+			if (firmware == NULL)
+				return FALSE;
+			fu_firmware_set_idx(firmware, fu_firmware_get_idx(FU_FIRMWARE(entry)));
+			g_ptr_array_add(files, g_steal_pointer(&firmware));
+		}
+	}
+
```

Comment:
what about cases that you load fwupdx64.efi as an arg?  Shouldn't that be parsed too?

Reply:
Yes, good catch -- fixed.